### DISCUSSION TITLE
Send the application closing event to Faro in React

### DISCUSSION THREAD

**Discussion by Ale142** [grafana/faro-web-sdk#588] [Q&A]
Hello everyone, from what I read from the [documentation](https://grafana.com/docs/grafana-cloud/monitor-applications/frontend-observability/architecture/instrumentations/#how-session-tracking-works), the user session has 3 lifecycle events:

- session_start
- session_resume
- session_extend

I was trying to create a custom event to report on closing of an application tab/window to send to Faro but It seems is not working and I need some advice from you. 

Here you have the Faro initialization in React:
```tsx
initializeFaro({
                isolate: true,
                instrumentations: [
                    ...getWebInstrumentations(),
                ],
                sessionTracking: {
                    enabled: true,
                    persistent: false,
                },
                url: "..."
                app: {
                    ...
                },

            });
```
In another component I have created a `useEffect` with this code that creates a listener for the [visibilitychange event](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event) that is the recommeded method to use.
```tsx
 useEffect(() => {
        window.addEventListener('visibilitychange', () => {
            // Push the event on unload of the window
            faroApi?.api.pushEvent('tab-closed');
        });
        // Clear listener
        return () => {
            window.removeEventListener('visibilitychange', () => {
                faroApi?.api.pushEvent('tab-closed');
            });
        };
    }, [faroApi?.api]);
``` 
But when I close the tab the method seems not to be fired. Do you have any advice?

**Comment by codecapitano** [comment]
Hi @Ale142 the code looks good.
Did you already double check if the effect callback has been called and if yes, if the faro api is available?

**Reply by codecapitano** [reply]
One thing to keep in mind.
The listener above will the called for each visibility change. 
It‘s recommended to check ˋdocument.visibilityStateˋ is ˋhiddenˋ before pushing the tab-closed event.

**Reply by Ale142** [reply]
Hi @codecapitano thanks for your answer and the advice.

As soon as I switch the tab the event is correctly fired (`tab-hidden`/`tab-visible` depending on the `visibilityState`, thus the callback on the `addEventListener` is called correctly). But when I close the tab the event is not fired. I tried to debug the behavior using the debug mode of the browser (Chrome) but I can't debug the time when the cleanup function should be called (during the unmounting of the component when the application tab is closed).

**Reply by codecapitano** [reply]
Hmm ok.
I think the browser already has unloaded the page before the listener is executed.

**Reply by codecapitano** [reply]
You can use the `beforeUnload` event but it is unreliable in some cases and can have an impact on back/forward cache.

**Reply by Ale142** [reply]
Yes that's the reason why I avoided to used the `beforeUnload` event. Anyway, do you know if is it planned to add a new [session lifecycle event ](https://grafana.com/docs/grafana-cloud/monitor-applications/frontend-observability/architecture/instrumentations/#how-session-tracking-works) that specifies when the session end by closing the tab/window?

**Reply by codecapitano** [reply]
No it's not planned atm.

**Reply by codecapitano** [reply]
Maybe try the `pagehide` event.
At least this plays nicely with bf cache.
Reliability issues on mobile still apply. 

* https://developer.mozilla.org/en-US/docs/Web/API/Window/pagehide_event
* https://developer.chrome.com/docs/web-platform/page-lifecycle-api
